File: src/Tests/Behavioral/ItabLateRegistration/latelib/latelib.cs
namespace go.ItabLateRegistration;

using lib = go.ItabLateRegistration.lib_package;
using go.ItabLateRegistration;

partial class latelib_package {

public static (@string, nint) Prime() {
    ref var c = ref heap<lib.Circle>(out var Ꮡc);
    c = new lib.Circle(R: 7);
    ref var q = ref heap<lib.Square>(out var Ꮡq);
    q = new lib.Square(S: 8);
    lib.Named n = new lib_CircleжNamed(Ꮡc);
    lib.Sized s = new lib_SquareжSized(Ꮡq);
    return (n.Name(), s.Size());
}

} // end latelib_package

File: src/Tests/Behavioral/ItabLateRegistration/latelib/latelib.go
// Package latelib exists to be loaded LATE. Its only job is to hold conversion sites the
// converter CAN record, so go2cs-gen emits a generated nominal adapter for each pair and
// registers it from this assembly's module initializer — which the CLR does not run until
// something in this module is first used.
package latelib

import "ItabLateRegistration/lib"

// Prime performs the recordable conversions. Assigning a POINTER to the interface is what makes
// the generated adapter register itself: a value-typed nominal match is satisfied by having the
// struct implement the interface directly, which leaves nothing to register at run time.
func Prime() (string, int) {
	c := lib.Circle{R: 7}
	q := lib.Square{S: 8}

	var n lib.Named = &c
	var s lib.Sized = &q

	return n.Name(), s.Size()
}
